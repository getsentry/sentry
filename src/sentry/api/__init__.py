from .analytics import *  # NOQA
from .client import ApiClient

client = ApiClient()
