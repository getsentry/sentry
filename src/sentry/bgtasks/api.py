from __future__ import absolute_import

import six
import time
import random
import logging
import threading
from contextlib import contextmanager

from django.conf import settings


logger = logging.getLogger("sentry.bgtasks")
tasks = {}


def bgtask(roles=None, interval=60):
    def decorator(f):
        return BgTask(callback=f, roles=roles, interval=interval)

    return decorator


class BgTask(object):
    def __init__(self, callback, roles=None, interval=60):
        self.callback = callback
        self.roles = roles or []
        self.interval = interval
        self.running = False

    @property
    def name(self):
        return "%s:%s" % (self.callback.__module__, self.callback.__name__)

    def run(self):
        if self.running:
            return
        self.running = True

        next_run = time.time() + self.interval * random.random()
        while self.running:
            now = time.time()
            if now >= next_run:
                try:
                    self.callback()
                except Exception:
                    logging.error("bgtask.failed", exc_info=True, extra=dict(task_name=self.name))
                next_run = now + self.interval
            time.sleep(1.0)

    def reconfigure(self, cfg):
        if "roles" in cfg:
            self.roles = cfg["roles"]
        if "interval" in cfg:
            self.interval = cfg["interval"]

    def spawn_daemon(self):
        if self.running:
            return
        logger.info("bgtask.spawn", extra=dict(task_name=self.name))
        t = threading.Thread(target=self.run)
        t.setDaemon(True)
        t.start()

    def stop(self):
        logger.info("bgtask.stop", extra=dict(task_name=self.name))
        self.running = False


def get_task(task_name):
    module, task_cls = task_name.split(":", 1)
    mod = __import__(module, None, None, [task_cls])
    return getattr(mod, task_cls)


def spawn_bgtasks(role):
    for import_name, cfg in six.iteritems(settings.BGTASKS):
        task = get_task(import_name)
        # This is already running
        if task.name in tasks:
            continue
        task.reconfigure(cfg)
        if role not in task.roles:
            continue
        task.spawn_daemon()
        tasks[task.name] = task


def shutdown_bgtasks():
    for task_name, task in list(six.iteritems(tasks)):
        task.stop()
        tasks.pop(task_name, None)


@contextmanager
def managed_bgtasks(role):
    spawn_bgtasks(role)
    try:
        yield
    finally:
        shutdown_bgtasks()
