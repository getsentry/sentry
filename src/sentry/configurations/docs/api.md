# Configurations API

Host: https://sentry.io/api/0

**Authors.**

@cmanallen

## Configuration [/projects/<organization_id_or_slug>/<project_id_or_slug>/configuration/]

### Get Configuration [GET]

Retrieve the project's configuration.

**Attributes**

| Column             | Type   | Description                     |
| ------------------ | ------ | ------------------------------- |
| id                 | number | Project id.                     |
| sample_rate        | number | A number value between 0 and 1. |
| traces_sample_rate | number | A number value between 0 and 1. |
| user_config        | any    | Arbitrary user supplied JSON.   |

- Response 200

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 1.0,
      "traces_sample_rate": 0.5,
      "user_config": {
        "hello": "world"
      }
    }
  }
  ```

### Update Configuration [PATCH]

Update the project's configuration.

- Request

  ```json
  {
    "data": {
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

- Response 201

  ```json
  {
    "data": {
      "id": 1,
      "sample_rate": 0.2,
      "traces_sample_rate": 0.5,
      "user_config": ["hello", "world"]
    }
  }
  ```

### Delete Configuration [DELETE]

Delete the project's configuration.

- Response 204
