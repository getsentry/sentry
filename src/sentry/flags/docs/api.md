# Flags API

Host: https://sentry.io/api/0

**Authors.**

@cmanallen

**How to read this document.**

This document is structured by resource with each resource having actions that can be performed against it. Every action that either accepts a request or returns a response WILL document the full interchange format. Clients may opt to restrict response data or provide a subset of the request data.

## Flag Logs [/organizations/<organization_id_or_slug>/flag-log/]

- Parameters
  - query (optional, string) - Search query with space-separated field/value pairs. ie: `?query=environment:prod AND project:3`.
  - start (optional, string) - ISO 8601 format (`YYYY-MM-DDTHH:mm:ss.sssZ`)
  - end (optional, string) - ISO 8601 format. Required if `start` is set.
  - statsPeriod (optional, string) - A positive integer suffixed with a unit type.
  - cursor (optional, string)`
  - per_page (optional, number)
    Default: 10
  - offset (optional, number)
    Default: 0

### Browse Flag Logs [GET]

Retrieve a collection of flag logs.

**Attributes**

| Column          | Type   | Description                                          |
| --------------- | ------ | ---------------------------------------------------- |
| action          | string | Enum of `created`, `updated`, or `deleted`.          |
| flag            | string | The name of the flag changed.                        |
| created_at      | string | ISO-8601 timestamp of when the flag was changed.     |
| created_by      | string | The user responsible for the change.                 |
| created_by_type | string | Enum of `email`, `id`, or `name`.                    |
| tags            | object | A collection of provider-specified scoping metadata. |

- Response 200

  ```json
  {
    "data": [
      {
        "action": "created",
        "flag": "my-flag-name",
        "created_at": "2024-01-01T05:12:33",
        "created_by": "2552",
        "created_by_type": "id",
        "tags": {
          "environment": "production"
        }
      }
    ]
  }
  ```

## Flag Log [/organizations/<organization_id_or_slug>/flag-log/<flag>/]

### Fetch Flag Log [GET]

Retrieve a single flag log instance.

- Response 200

  ```json
  {
    "data": {
      "action": "updated",
      "flag": "new-flag-name",
      "created_at": "2024-11-19T19:12:55",
      "created_by": "user@site.com",
      "created_by_type": "email",
      "tags": {
        "environment": "development"
      }
    }
  }
  ```

## Webhooks [/webhooks/flags/organization/<organization_id_or_slug>/provider/<provider>/]

### Create Flag Log [POST]

The shape of the request object varies by provider. The `<provider>` URI parameter informs the server of the shape of the request and it is on the server to handle the provider. The following providers are supported: Unleash, Split, and LaunchDarkly.

**Flag Pole Example:**

Flag pole is Sentry owned. It matches our audit-log resource because it is designed for that purpose.

- Request (application/json)

  ```json
  {
    "data": [
      {
        "action": "updated",
        "flag": "flag-name",
        "created_at": "2024-11-19T19:12:55",
        "created_by": "colton.allen@sentry.io",
        "tags": {
          "commit_sha": "1f33a107d7cd060ab9c98e11c9e5a62dc1347861"
        }
      }
    ]
  }
  ```

- Response 201
