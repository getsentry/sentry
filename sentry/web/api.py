"""
sentry.web.views
~~~~~~~~~~~~~~~~

:copyright: (c) 2010 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
import logging

from django.http import HttpResponse, HttpResponseBadRequest, \
  HttpResponseForbidden
from django.views.decorators.csrf import csrf_exempt
from django.views.decorators.http import require_http_methods

from sentry.conf import settings
from sentry.coreapi import project_from_auth_vars, project_from_api_key_and_id, \
  project_from_id, decode_and_decompress_data, safely_load_json_string, \
  validate_data, insert_data_to_database, APIError, APIUnauthorized, \
  extract_auth_vars, InvalidTimestamp
from sentry.models import Group, GroupBookmark, Project, View, ProjectDomain
from sentry.utils import json
from sentry.utils.http import is_same_domain, apply_access_control_headers
from sentry.web.decorators import has_access
from sentry.web.frontend.groups import _get_group_list
from sentry.web.helpers import render_to_response, render_to_string

logger = logging.getLogger('sentry.errors.coreapi')


@csrf_exempt
@require_http_methods(['POST', 'OPTIONS'])
def store(request):
    """
    The primary endpoint for storing new events.

    This will validate the client's authentication and data, and if
    successfull pass on the payload to the internal database handler.

    Authentication works in three flavors:

    1. Explicit signed requests

       These are implemented using the documented signed request protocol, and
       require an authentication header which is signed using with the project
       member's secret key.

    2. Explicit trusted requests

       Generally used for communications with client-side platforms (such as
       JavaScript in the browser), they require the GET variables public_key
       and project_id, as well as an HTTP_REFERER to be set from a trusted
       domain.

    3. Implicit trusted requests

       Used by the Sentry core, they are only available from same-domain requests
       and do not require any authentication information. They only require that
       the user be authenticated, and a project_id be sent in the GET variables.

    """
    try:
        if request.method == 'POST':
            auth_vars = extract_auth_vars(request)
            data = request.raw_post_data

            if auth_vars:
                server_version = auth_vars.get('sentry_version', '1.0')
                client = auth_vars.get('sentry_client')
            else:
                server_version = request.GET.get('version', '1.0')
                client = request.META.get('HTTP_USER_AGENT', request.GET.get('client'))

            if server_version not in ('1.0', '2.0'):
                raise APIError('Client/server version mismatch: Unsupported version: %r' % server_version)

            if server_version != '1.0' and not client:
                raise APIError('Client request error: Missing client version identifier.')

            referrer = request.META.get('HTTP_REFERER')

            if auth_vars:
                project = project_from_auth_vars(auth_vars, data)
            elif request.GET.get('api_key') and request.GET.get('project_id'):
                # public requests only need referrer validation for CSRF
                project = project_from_api_key_and_id(request.GET['api_key'], request.GET['project_id'])
                if not ProjectDomain.test(project, referrer):
                    raise APIUnauthorized()
            elif request.GET.get('project_id') and request.user.is_authenticated() and \
                 is_same_domain(request.build_absolute_uri(), referrer):
                # authenticated users are simply trusted to provide the right id
                project = project_from_id(request)
            else:
                raise APIUnauthorized()

            if not data.startswith('{'):
                data = decode_and_decompress_data(data)
            data = safely_load_json_string(data)

            try:
                validate_data(project, data)
            except InvalidTimestamp:
                # Log the error, remove the timestamp, and revalidate
                logger.error('Client %r passed an invalid value for timestamp %r' % (
                    data['timestamp'],
                    client or '<unknown client>',
                ))
                del data['timestamp']
                validate_data(project, data)

            insert_data_to_database(data)
    except APIError, error:
        response = HttpResponse(error.msg, status=error.http_status)
    else:
        response = HttpResponse('')
    return apply_access_control_headers(response)


@csrf_exempt
@has_access
def notification(request, project):
    return render_to_response('sentry/partial/_notification.html', request.GET)


@csrf_exempt
@has_access
def poll(request, project):
    from sentry.templatetags.sentry_helpers import as_bookmarks, handle_before_events

    offset = 0
    limit = settings.MESSAGES_PER_PAGE

    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    filters, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list = list(event_list[offset:limit])
    handle_before_events(request, event_list)

    data = [
        {
            'id': m.pk,
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
                'is_bookmarked': b,
            }).strip(),
            'title': m.message_top(),
            'message': m.error(),
            'level': m.get_level_display(),
            'logger': m.logger,
            'count': m.times_seen,
            'score': getattr(m, 'sort_value', None),
        } for m, b in as_bookmarks(event_list, request.user)]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def resolve(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()
    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    Group.objects.filter(pk=group.pk).update(status=1)
    group.status = 1

    data = [
        (m.pk, {
            'html': render_to_string('sentry/partial/_group.html', {
                'group': m,
                'request': request,
            }).strip(),
            'count': m.times_seen,
        }) for m in [group]]

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def bookmark(request, project):
    gid = request.REQUEST.get('gid')
    if not gid:
        return HttpResponseForbidden()

    if not request.user.is_authenticated():
        return HttpResponseForbidden()

    try:
        group = Group.objects.get(pk=gid)
    except Group.DoesNotExist:
        return HttpResponseForbidden()

    gb, created = GroupBookmark.objects.get_or_create(
        project=group.project,
        user=request.user,
        group=group,
    )
    if not created:
        gb.delete()

    response = HttpResponse(json.dumps({'bookmarked': created}))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def clear(request, project):
    view_id = request.GET.get('view_id')
    if view_id:
        try:
            view = View.objects.get(pk=view_id)
        except View.DoesNotExist:
            return HttpResponseBadRequest()
    else:
        view = None

    _, event_list = _get_group_list(
        request=request,
        project=project,
        view=view,
    )

    event_list.update(status=1)

    data = []
    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response


@csrf_exempt
@has_access
def chart(request, project):
    gid = request.REQUEST.get('gid')
    days = int(request.REQUEST.get('days', '90'))

    if gid:
        try:
            group = Group.objects.get(pk=gid)
        except Group.DoesNotExist:
            return HttpResponseForbidden()

        data = Group.objects.get_chart_data(group, max_days=days)
    else:
        data = Project.objects.get_chart_data(project, max_days=days)

    response = HttpResponse(json.dumps(data))
    response['Content-Type'] = 'application/json'
    return response
