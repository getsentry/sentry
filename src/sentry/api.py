"""
sentry.api
~~~~~~~~~~

:copyright: (c) 2010-2012 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""

import base64
from django.contrib.auth.models import User, AnonymousUser
from tastypie import fields
from tastypie.authentication import Authentication
from tastypie.authorization import Authorization
from tastypie.api import Api
from tastypie.resources import ModelResource
from sentry.models import Project, Team, TeamMember, Group, Event, ProjectKey
from sentry.web.helpers import get_project_list


class SentryAuthentication(Authentication):
    def is_authenticated(self, request, **kwargs):
        """
        Simple basic auth where username is the public key, and password
        is the secret key.
        """
        auth = request.META.get('HTTP_AUTHORIZATION')
        if not auth:
            return False

        if not auth.startswith('Basic '):
            return

        try:
            decoded = base64.b64decode(auth.split(' ', 1)[-1])
            public, secret = decoded.split(':', 1)
        except Exception:
            return False

        try:
            apikey = ProjectKey.objects.get(public_key=public, secret_key=secret)
        except ProjectKey.DoesNotExist:
            return False

        request.projects = [apikey.project] or get_project_list(apikey.user)
        if apikey.user:
            request.teams = Team.objects.get_for_user(apikey.user)
        else:
            request.teams = []
        request.user = apikey.user or AnonymousUser()
        request.apikey = apikey

        return True

    def get_identifier(self, request):
        # XXX: no idea what this is for
        return request.apikey.id


class TeamAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(pk__in=[t.pk for t in request.teams])
        return object_list.none()


class TeamBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(team__project__in=request.projects)
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(pk__in=[p.id for p in request.projects])
        elif request.teams:
            return object_list.filter(team__in=request.teams)
        return object_list.none()


class ProjectBasedAuthorization(Authorization):
    def apply_limits(self, request, object_list):
        if request.projects:
            return object_list.filter(project__in=request.projects)
        elif request.teams:
            return object_list.filter(project__team__in=request.teams)
        return object_list.none()


class UserResource(ModelResource):
    username = fields.CharField('username', readonly=True)
    # name = fields.CharField('first_name', readonly=True)

    class Meta:
        queryset = User.objects.all()
        detail_uri_name = 'username'
        include_resource_uri = False
        resource_name = 'user'
        detail_allowed_methods = []
        list_allowed_methods = []
        fields = ['username']
        authentication = SentryAuthentication()


class TeamResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    members = fields.ToManyField('sentry.api.TeamMemberResource', 'member_set', full=True)

    class Meta:
        queryset = Team.objects.all()
        detail_uri_name = 'slug'
        resource_name = 'team'
        detail_allowed_methods = ['get']
        list_allowed_methods = ['get']
        fields = ['id', 'name']
        authentication = SentryAuthentication()
        authorization = TeamAuthorization()


class TeamMemberResource(ModelResource):
    user = fields.ToOneField(UserResource, 'user', full=True)
    # access = fields.CharField('type')

    class Meta:
        queryset = TeamMember.objects.all()
        include_resource_uri = False
        detail_allowed_methods = ['get']
        list_allowed_methods = []
        fields = ['user']
        authentication = SentryAuthentication()
        authorization = TeamBasedAuthorization()

    # def dehydrate_access(self, bundle):
    #     return bundle.obj.get_type_display()


class ProjectResource(ModelResource):
    slug = fields.CharField(attribute="slug", readonly=True)
    owner = fields.ToOneField(UserResource, 'owner', full=True, readonly=True)
    team = fields.ToOneField(TeamResource, 'team', full=True)

    class Meta:
        queryset = Project.objects.filter(status=0)
        detail_uri_name = 'slug'
        resource_name = 'project'
        fields = ['id', 'name']
        allowed_methods = ['get']
        authentication = SentryAuthentication()
        authorization = ProjectAuthorization()


class GroupResource(ModelResource):
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    team = fields.ToOneField(TeamResource, 'team', readonly=True)
    message = fields.CharField('message', readonly=True)
    times_seen = fields.IntegerField('times_seen', readonly=True)
    last_seen = fields.DateTimeField('last_seen', readonly=True)
    first_seen = fields.DateTimeField('first_seen', readonly=True)
    latest_event = fields.ToOneField('sentry.api.EventResource', 'latest_event', readonly=True, full=True)
    # events = fields.ToManyField('sentry.api.EventResource', 'event_set')

    class Meta:
        queryset = Group.objects.all()
        resource_name = 'group'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    def full_dehydrate(self, bundle):
        # HACK:
        bundle.obj.team = bundle.obj.project.team
        bundle.obj.latest_event = bundle.obj.get_latest_event()
        return super(GroupResource, self).full_dehydrate(bundle)


class EventResource(ModelResource):
    group = fields.ToOneField(GroupResource, 'group', readonly=True)
    project = fields.ToOneField(ProjectResource, 'project', readonly=True)
    message = fields.CharField('message', readonly=True)
    # data = fields.DictField(readonly=True)
    datetime = fields.DateTimeField('datetime', readonly=True)

    class Meta:
        queryset = Event.objects.all()
        resource_name = 'event'
        allowed_methods = ['get']
        fields = ['id']
        authentication = SentryAuthentication()
        authorization = ProjectBasedAuthorization()

    # def dehydrate_data(self, bundle):
    #     return bundle.obj.data


v1_api = Api(api_name='v1')
v1_api.register(TeamResource())
v1_api.register(TeamMemberResource())
v1_api.register(ProjectResource())
v1_api.register(GroupResource())
v1_api.register(EventResource())
