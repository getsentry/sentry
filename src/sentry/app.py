"""
sentry.app
~~~~~~~~~~

:copyright: (c) 2010-2014 by the Sentry Team, see AUTHORS for more details.
:license: BSD, see LICENSE for more details.
"""
from __future__ import absolute_import

from threading import local

from django.conf import settings
from raven.contrib.django.models import client

from sentry.utils import redis
from sentry.utils.imports import import_string
from sentry.utils.locking.backends.redis import RedisLockBackend
from sentry.utils.locking.manager import LockManager


class State(local):
    request = None
    data = {}

env = State()


def get_instance(path, options):
    cls = import_string(path)
    return cls(**options)


# TODO(dcramer): this is getting heavy, we should find a better way to structure
# this
buffer = get_instance(settings.SENTRY_BUFFER, settings.SENTRY_BUFFER_OPTIONS)
digests = get_instance(settings.SENTRY_DIGESTS, settings.SENTRY_DIGESTS_OPTIONS)
quotas = get_instance(settings.SENTRY_QUOTAS, settings.SENTRY_QUOTA_OPTIONS)
nodestore = get_instance(
    settings.SENTRY_NODESTORE, settings.SENTRY_NODESTORE_OPTIONS)
ratelimiter = get_instance(
    settings.SENTRY_RATELIMITER, settings.SENTRY_RATELIMITER_OPTIONS)
search = get_instance(settings.SENTRY_SEARCH, settings.SENTRY_SEARCH_OPTIONS)
tsdb = get_instance(settings.SENTRY_TSDB, settings.SENTRY_TSDB_OPTIONS)
raven = client
locks = LockManager(RedisLockBackend(redis.clusters.get('default')))
