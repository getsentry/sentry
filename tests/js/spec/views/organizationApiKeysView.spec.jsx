import React from 'react';
import {mount} from 'enzyme';

import {Client} from 'app/api';
import OrganizationApiKeysView
  from 'app/views/settings/organization/apiKeys/organizationApiKeysView';

const childContextTypes = {
  organization: React.PropTypes.object,
  router: React.PropTypes.object,
  location: React.PropTypes.object
};

describe('OrganizationApiKeysView', function() {
  beforeEach(function() {
    Client.clearMockResponses();
    Client.addMockResponse({
      url: '/organizations/org-slug/api-keys/',
      method: 'GET',
      body: [TestStubs.ApiKey()]
    });
    Client.addMockResponse({
      url: '/organizations/org-slug/api-keys/1/',
      method: 'GET',
      body: TestStubs.ApiKey()
    });
  });

  it('renders', function() {
    let wrapper = mount(<OrganizationApiKeysView params={{orgId: 'org-slug'}} />, {
      context: {
        router: TestStubs.router(),
        organization: TestStubs.Organization(),
        location: TestStubs.location()
      },
      childContextTypes
    });
    expect(wrapper.state('loading')).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('can delete a key', function() {
    let wrapper = mount(<OrganizationApiKeysView params={{orgId: 'org-slug'}} />, {
      context: {
        router: TestStubs.router(),
        organization: TestStubs.Organization(),
        location: TestStubs.location()
      },
      childContextTypes
    });
    OrganizationApiKeysView.handleRemove = jest.fn();
    expect(OrganizationApiKeysView.handleRemove).not.toHaveBeenCalled();

    // Click remove button
    wrapper.find('.icon-trash').simulate('click');
    wrapper.update();

    // expect a modal
    let modal = wrapper.find('Modal');
    expect(modal.first().prop('show')).toBe(true);

    // TODO
    // wrapper.find('Modal').last().find('Button').last().simulate('click');

    // expect(OrganizationApiKeysView.handleRemove).toHaveBeenCalled();
  });
});
