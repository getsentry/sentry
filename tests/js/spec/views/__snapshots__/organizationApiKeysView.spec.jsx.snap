// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationApiKeysView renders 1`] = `
<OrganizationApiKeysView
  params={
    Object {
      "orgId": "org-slug",
    }
  }
>
  <DocumentTitle
    title="Organization Name API Keys"
  >
    <div>
      <SpreadLayout
        center={true}
        className="page-header"
        responsive={false}
      >
        <div
          className="spread-layout page-header center"
        >
          <h3>
            Api Keys
          </h3>
          <Button
            disabled={false}
            onClick={[Function]}
            priority="primary"
            type="button"
          >
            <button
              className="button button-primary"
              disabled={false}
              onClick={[Function]}
              role="button"
              type="button"
            >
              <FlowLayout
                truncate={false}
              >
                <div
                  className="flow-layout"
                >
                  <span
                    className="button-label"
                  >
                    New API Key
                  </span>
                </div>
              </FlowLayout>
            </button>
          </Button>
        </div>
      </SpreadLayout>
      <p>
        API keys grant access to the
         
        <a
          href="https://docs.sentry.io/hosted/api/"
          rel="nofollow"
          target="_blank"
        >
          developer web API
        </a>
        . If you're looking to configure a Sentry client, you'll need a client key which is available in your project settings.
      </p>
      <div
        className="alert alert-block alert-info"
      >
        psst. Until Sentry supports OAuth, you might want to switch to using
         
        <Link
          to="/api/"
        >
          <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/api/"
          >
            <a
              onClick={[Function]}
              style={Object {}}
            >
              Auth Tokens
            </a>
          </Link>
        </Link>
         
        instead.
      </div>
      <table
        className="table api-key-list"
      >
        <colgroup>
          <col
            style={
              Object {
                "width": "40%",
              }
            }
          />
          <col
            style={
              Object {
                "width": "40%",
              }
            }
          />
          <col
            style={
              Object {
                "width": "20%",
              }
            }
          />
        </colgroup>
        <tbody>
          <tr>
            <td>
              <h5>
                <Link
                  to="/organizations/org-slug/api-keys/1"
                >
                  <Link
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="/organizations/org-slug/api-keys/1"
                  >
                    <a
                      onClick={[Function]}
                      style={Object {}}
                    >
                      Default
                    </a>
                  </Link>
                </Link>
              </h5>
            </td>
            <td>
              <div
                className="form-control disabled auto-select"
              >
                aa624bcc12024702a202cd90be5feda0
              </div>
            </td>
            <td
              className="align-right"
            >
              <Link
                className="btn btn-default btn-sm"
                style={
                  Object {
                    "marginRight": 4,
                  }
                }
                to="/organizations/org-slug/api-keys/1"
              >
                <Link
                  className="btn btn-default btn-sm"
                  onlyActiveOnIndex={false}
                  style={
                    Object {
                      "marginRight": 4,
                    }
                  }
                  to="/organizations/org-slug/api-keys/1"
                >
                  <a
                    className="btn btn-default btn-sm"
                    onClick={[Function]}
                    style={
                      Object {
                        "marginRight": 4,
                      }
                    }
                  >
                    Edit Key
                  </a>
                </Link>
              </Link>
              <LinkWithConfirmation
                className="btn btn-default btn-sm"
                message="Are you sure you want to remove this API key?"
                onConfirm={[Function]}
                title="Remove API Key?"
              >
                <Confirm
                  cancelText="Cancel"
                  confirmText="Confirm"
                  message="Are you sure you want to remove this API key?"
                  onConfirm={[Function]}
                  priority="primary"
                >
                  <span>
                    <a
                      className="btn btn-default btn-sm"
                      onClick={[Function]}
                      title="Remove API Key?"
                    >
                      <span
                        className="icon-trash"
                      />
                    </a>
                    <Modal
                      animation={false}
                      autoFocus={true}
                      backdrop={true}
                      bsClass="modal"
                      dialogComponentClass={[Function]}
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        ModalManager {
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "modals": Array [],
                        }
                      }
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      restoreFocus={true}
                      show={false}
                    >
                      <Modal
                        autoFocus={true}
                        backdrop={true}
                        backdropClassName="modal-backdrop"
                        backdropTransitionTimeout={150}
                        containerClassName="modal-open"
                        dialogTransitionTimeout={300}
                        enforceFocus={true}
                        keyboard={true}
                        manager={
                          ModalManager {
                            "containers": Array [],
                            "data": Array [],
                            "handleContainerOverflow": true,
                            "hideSiblingNodes": true,
                            "modals": Array [],
                          }
                        }
                        onEntering={[Function]}
                        onExited={[Function]}
                        onHide={[Function]}
                        renderBackdrop={[Function]}
                        restoreFocus={true}
                        show={false}
                      />
                    </Modal>
                  </span>
                </Confirm>
              </LinkWithConfirmation>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </DocumentTitle>
</OrganizationApiKeysView>
`;
